### PR TITLE
sotw: Ordered ADS mode

### PR DESCRIPTION
**Pull request by alecholmez**
introduce optional resource ordering for guaranteed eventual consistency in ADS mode.

### CODE REVIEW

**Comment by alecholmez** [comment]
This PR branches off the work found here: https://github.com/envoyproxy/go-control-plane/pull/461 and brings it up to speed with the current state of go-control-plane. We've found quite a few resource ordering issues in upstream xDS conformance test suites and this will hopefully provide a remedy to some of those problems. 

I tried to introduce the change in a non intrusive way, if no `opts` are provided to the server everything will execute as before. If the new opt `sotw.WithOrderedADS()` is passed to a server, the resources will be in guaranteed ordering as in the prior art PR. This provides guaranteed eventual consistency which we did not have in the previous implementation but gives up some of the pressure relief from immediate response sending.

File: pkg/integration/ttl_integration_test.go [hunk 1]
```diff
@@ -31,14 +31,11 @@ func (log logger) Warnf(format string, args ...interface{})  { log.t.Logf(format
 func (log logger) Errorf(format string, args ...interface{}) { log.t.Logf(format, args...) }
 
 func TestTTLResponse(t *testing.T) {
-
```
Comment (alecholmez) [inline pkg/integration/ttl_integration_test.go:34]: Ran this file through a formatter hence these changes.

**Comment by alecholmez** [comment]
Currently investigating the integration test failure.... seems to me that the server gets stuck somewhere and disconnects

File: pkg/cache/types/types.go [hunk 2]
```diff
@@ -35,8 +35,8 @@ type ResponseType int
 const (
 	Endpoint ResponseType = iota
 	Cluster
-	Route
 	ScopedRoute
+	Route
```
Comment (sunjayBhatia) [inline pkg/cache/types/types.go:39]: this is nitpicky but could this technically be a breaking change? in case someone had comparisons on this type, seems unfortunate but maybe

Comment (sunjayBhatia) [inline pkg/cache/types/types.go:39]: this is nitpicky but could this technically be a breaking change? in case someone had comparisons on this type, seems unfortunate but maybe

File: pkg/cache/types/types.go [hunk 3]
```diff
@@ -35,8 +35,8 @@ type ResponseType int
 const (
 	Endpoint ResponseType = iota
 	Cluster
```
Comment (sunjayBhatia) [inline pkg/cache/types/types.go:37]: should `Cluster` sort before `Endpoint`?

Comment (sunjayBhatia) [inline pkg/cache/types/types.go:37]: should `Cluster` sort before `Endpoint`?

File: pkg/cache/v3/order_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		{
+			ID:      5,
+			TypeURL: resource.ScopedRouteType,
+		},
+	}
+
+	orderedKeys := unorderedKeys
+	sort.Sort(orderedKeys)
+
+	if !sort.IsSorted(orderedKeys) {
```
Comment (sunjayBhatia) [inline pkg/cache/v3/order_test.go:38]: i dont think this really works to test it this way, its just checking itself with itself, right now I get this order when i print `orderedKeys` and the test passes:

```
{4 type.googleapis.com/envoy.config.listener.v3.Listener}
{3 type.googleapis.com/envoy.config.route.v3.RouteConfiguration}
{5 type.googleapis.com/envoy.config.route.v3.ScopedRouteConfiguration}
{2 type.googleapis.com/envoy.config.cluster.v3.Cluster}
{1 type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment}
```

File: pkg/cache/types/types.go (hunk 3 above)
Comment (alecholmez) [reply pkg/cache/types/types.go:37]: ^ this I have no idea, I had it in the original order and that was causing the bug that @maplebeats was seeing

File: pkg/cache/types/types.go (hunk 2 above)
Comment (alecholmez) [reply pkg/cache/types/types.go:39]: Ah yea that's a good point... hm

File: pkg/cache/v3/order_test.go (hunk 4 above)
Comment (alecholmez) [reply pkg/cache/v3/order_test.go:38]: Yea so this test is pretty much pointless. It was really just to show that we can sort the keys in the order they're expected, LDS before RDS and SRDS, and CDS before EDS. I guess LDS before CDS doesn't matter though

File: pkg/cache/types/types.go (hunk 3 above)
Comment (maplebeats) [reply pkg/cache/types/types.go:37]: cluster should send before endpoints in cache server, otherwise cds will keep waiting for eds response.

**Comment by alecholmez** [comment]
Haven’t had much bandwidth to work this recently will get back to it soon.

File: pkg/cache/types/types.go (hunk 2 above)
Comment (alecholmez) [reply pkg/cache/types/types.go:39]: I think the solution to this imo is just bump the released version. We can note this in the changelog but it's a critical part of the ordering problem.

File: pkg/cache/v3/order_test.go (hunk 4 above)
Comment (alecholmez) [reply pkg/cache/v3/order_test.go:38]: Also want to note here that when I started sending scoped routes before routes envoy began to complain. I wasn't sure if that was a bug so I moved it back and all seemed to be okay. I think the most important ordering here is that LDS and CDS send first so envoy can build its config tree properly.

**Comment by marcosrmendezthd** [comment]
any progress on this?

**Comment by alecholmez** [comment]
Need to fix the integration tests and then this should be ready for a rereview

**Comment by haoruolei** [comment]
Hi do we have updates on this PR?

**Comment by alecholmez** [comment]
Tests pass now just working through lint issue

File: pkg/cache/v3/status.go [hunk 5]
```diff
@@ -155,3 +159,26 @@ func (info *statusInfo) setDeltaResponseWatch(id int64, drw DeltaResponseWatch)
...
+		info.orderedWatches[index] = key{
+			ID:      id,
+			TypeURL: watch.Request.TypeUrl,
+		}
+		index++
+	}
+
+	// Sort our list which we can use in the SetSnapshot functions.
+	// This is only run when we enable ADS on the cache.
+	if order {
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/cache/v3/status.go:181]: One of the original algorithm designs was to allow 2 modes in ADS:

- A slower ordered mode
- A faster unordered mode

The latter could be helpful in scenarios that the management server needs the least possible back pressure. I just followed the original algorithm but I think there is merit allowing users to pick the highest-bandwidth mode of operation.

File: pkg/cache/v3/status.go [hunk 6]
```diff
@@ -155,3 +159,26 @@ func (info *statusInfo) setDeltaResponseWatch(id int64, drw DeltaResponseWatch)
 	defer info.mu.Unlock()
 	info.deltaWatches[id] = drw
 }
+
+// orderResponseWatches will track a list of watch keys and order them if
+// true is passed.
+func (info *statusInfo) orderResponseWatches(order bool) {
+	info.orderedWatches = make(keys, len(info.watches))
+
+	// This runs in O(n) which could become problematic when we have an extremely high watch count.
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/cache/v3/status.go:168]: You're right the sort is really fast but building that indexed ordered list is worse case scenario `O(len(info.watches))`. Luckily inserting into a map is `O(1)`. Honestly I can probably delete this comment because this operation really isn't that expensive.

File: pkg/cache/v3/status.go [hunk 7]
```diff
@@ -155,3 +159,26 @@ func (info *statusInfo) setDeltaResponseWatch(id int64, drw DeltaResponseWatch)
...
+	// TODO(alec): revisit this and optimize for speed.
+	var index int
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/cache/v3/status.go:170]: I tried to do that as an optimization. Append is kind of slow under the hood and pre-allocating/using the index is slightly faster.

I can undo that though if that's impacting readability.

File: pkg/cache/v3/simple.go [hunk 8]
```diff
@@ -232,20 +232,41 @@ func (cache *snapshotCache) SetSnapshot(ctx context.Context, node string, snapsh
...
 				delete(info.watches, id)
 			}
+			return nil
+		}
+
+		// If ADS is enabled we need to order response watches so we guarantee
+		// sending them in the correct order. Go's default implementation
+		// of maps are randomized order when ranged over.
+		if cache.ads {
+			info.orderResponseWatches(cache.ads)
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/cache/v3/simple.go:256]: Nice catch thank you

File: pkg/server/sotw/v3/ads.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+			// Use the multiplexed channel for new watches.
+			responder := sw.watches.responders[resource.AnyType].response
+			if w, ok := sw.watches.responders[typeURL]; ok {
+				// We've found a pre-existing watch, lets check and update if needed.
+				// If these requirements aren't satisfied, leave an open watch.
+				if w.nonce == "" || w.nonce == nonce {
+					w.close()
+
+					// Only process if we have an existing watch otherwise go ahead and create.
+					if err := processAllExcept(typeURL); err != nil {
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/ads.go:116]: @valerian-roche good call I'll drop a comment there but the original algorithm as outlined here:

https://github.com/envoyproxy/go-control-plane/pull/461

[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/ads.go:116]: It's been quite a long time since I've ported that logic over but if I remember correctly we don't actually drop the responses, they just get requeued and eventually are sent once the enforced order has been resolved. (CDS before EDS, LDS before RDS, etc...)

**Comment by alecholmez** [comment]
@arkodg Yup we'll need to port this over

File: pkg/server/sotw/v3/ads.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (valerian-roche) [reply pkg/server/sotw/v3/ads.go:116]: From what I remember, here it is dequeuing all pre-existing responses in the channel. If of the type it seems to simply be discarded. I'm afraid of the impact if the version in the request is not changed, as the control-plane will consider there is no change and not send new resources

[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/ads.go:118]: Hm maybe we can write a test to cover that

[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/ads.go:118]: Quick update, the tests actually do cover partially sent snapshots!

File: pkg/server/sotw/v3/server.go [hunk 10]
```diff
@@ -61,177 +72,78 @@ type server struct {
...
-		lastResponse := lastDiscoveryResponse{
-			nonce:     out.Nonce,
-			resources: make(map[string]struct{}),
-		}
-		for _, r := range resp.GetRequest().ResourceNames {
-			lastResponse.resources[r] = struct{}{}
-		}
-		lastDiscoveryResponses[resp.GetRequest().TypeUrl] = lastResponse
+	// increment nonce and convert it to base10
+	out.Nonce = strconv.FormatInt(atomic.AddInt64(&s.nonce, 1), 10)
```
Comment (valerian-roche) [inline pkg/server/sotw/v3/server.go:111]: Minor: do we need this atomic increase here? The rest of this method is not synchronized anyway

**Review by valerian-roche — APPROVED** [review]

File: go.mod [hunk 11]
```diff
@@ -1,6 +1,6 @@
 module github.com/envoyproxy/go-control-plane
 
-go 1.17
+go 1.19
```
Comment (jpeach) [inline go.mod:3]: Can we make this change in a separate PR, since it affects downstream projects?

File: pkg/cache/types/types.go [hunk 12]
```diff
@@ -41,8 +41,8 @@ type ResponseType int
 const (
```
Comment (jpeach) [inline pkg/cache/types/types.go:41]: Let's add a big fat comment to tell people that the ordering of these enum values matters, linking to the ADS reference if possible.

File: pkg/cache/v3/order_test.go (hunk 4 above)
Comment (jpeach) [inline pkg/cache/v3/order_test.go:38]:
```suggestion
assert.True(t, sort.IsSorted(orderedKeys))
assert.NotEqual(t, unorderedKeys, orderedKeys)
```

This shows that the sorting does something, then you can add a canonically sorted expected value as well.

File: pkg/log/log_test.go [hunk 13]
```diff
@@ -22,13 +22,11 @@ import (
...
-
 	xdsLogger := LoggerFuncs{
-		DebugFunc: logger.Printf,
-		InfoFunc:  logger.Printf,
-		WarnFunc:  logger.Printf,
-		ErrorFunc: logger.Printf,
+		DebugFunc: log.Printf,
+		InfoFunc:  log.Printf,
+		WarnFunc:  log.Printf,
+		ErrorFunc: log.Printf,
```
Comment (jpeach) [inline pkg/log/log_test.go:29]: Why do we need this change? Can we make a separate PR?

File: pkg/integration/ttl_integration_test.go (hunk 1 above)
Comment (jpeach) [reply pkg/integration/ttl_integration_test.go:35]: Can we make a separate PR with this, and just land it?

File: pkg/server/config/config.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package config
+
+// Opts for individual xDS implementations that can be
+// utilized through the functional opts pattern
```
Comment (jpeach) [inline pkg/server/config/config.go:4]: general nit: check that doc comments are correctly capitalized and punctuated.

File: pkg/server/config/config.go [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+type Opts struct {
+	// If true respond to ADS requests with a guaranteed resource ordering
+	Ordered bool
```
Comment (jpeach) [inline pkg/server/config/config.go:7]: TBH, I haven't looked at ordering for a long time. My recollection was that for ADS, the client was supposed to make requests in the order that was correct for it? If that's the case, do we really need to order anything in the server?

My other query is that, if ordering is a correctess issue, then shouldn't we be correct by default?

File: pkg/server/sotw/v3/ads.go [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+				}
+			}
+
+			typeURL := req.GetTypeUrl()
+			// Use the multiplexed channel for new watches.
+			responder := sw.watches.responders[resource.AnyType].response
+			if w, ok := sw.watches.responders[typeURL]; ok {
+				// We've found a pre-existing watch, lets check and update if needed.
+				// If these requirements aren't satisfied, leave an open watch.
+				if w.nonce == "" || w.nonce == nonce {
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/ads.go:114]: Um that's a good question, I left this as the previous behavior but maybe worth an issue?

File: pkg/server/sotw/v3/server.go (hunk 10 above)
Comment (alecholmez) [reply pkg/server/sotw/v3/server.go:111]: I think we do need this since N number of "sends" can increment the nonce counter which is shared. I could be wrong though I ported that from the previous code

File: pkg/server/sotw/v3/xds.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+				// When using ADS we need to order responses.
+				// This is guaranteed in the xDS protocol specification
+				// as ADS is required to be eventually consistent.
+				// More details can be found here if interested:
+				// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#eventual-consistency-considerations
+				if s.opts.Ordered {
+					// send our first request on the stream again so it doesn't get
+					// lost in processing on the new control loop
+					go func() {
+						reqCh <- req
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/xds.go:97]: Yea I did think about that. I couldn't really come up with anything smarter than this since we need to shuttle the request down into the processors, would be open to suggestions! If this does cause a bug we can patch after the fact too

File: pkg/server/stream/v3/stream.go [hunk 18]
```diff
@@ -71,56 +93,54 @@ func (s *StreamState) WatchesResources(resourceNames map[string]struct{}) bool {
...
-}
-
+// IsWildcard returns whether or not an xDS client requested in wildcard mode on the initial request.
 func (s *StreamState) IsWildcard() bool {
 	return s.wildcard
 }
 
+// IsOrdered returns whether or not the current stream should run as an ordered ADS stream.
+// This means less back-pressure relief but a guarantee of correct discovery response order.
+func (s *StreamState) IsOrdered(ordered bool) bool {
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/stream/v3/stream.go:124]: Ah yea I think I messed that up. I'll fix

File: go.mod (hunk 11 above)
Comment (alecholmez) [reply go.mod:3]: Yup I actually just opened this:
https://github.com/envoyproxy/go-control-plane/pull/714

File: pkg/log/log_test.go (hunk 13 above)
Comment (alecholmez) [reply pkg/log/log_test.go:29]: I think the linter was complaining about this? Honestly can't remember haha

File: pkg/server/config/config.go (hunk 15 above)
Comment (alecholmez) [reply pkg/server/config/config.go:7]: That's what I thought, the bug this closes out shows that envoy sometimes requests things out of order even in ADS. 

IMO for ADS we always want to be ordered so yes. If I recall correctly when a user configures the server to run in ADS mode it'll be ordered by default (but you can override that).

File: go.mod (hunk 11 above)
Comment (alecholmez) [reply go.mod]: Reverted back to 1.17 for this PR so we can get it in before the multi-module release.

File: pkg/server/sotw/v3/ads.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (valerian-roche) [reply pkg/server/sotw/v3/ads.go:117]: Yes I think an issue would be good to avoid missing this

File: pkg/server/sotw/v3/server.go (hunk 10 above)
Comment (valerian-roche) [reply pkg/server/sotw/v3/server.go:111]: My understanding is that all the "sends" run in the same goroutine on a stream, so I do not expect the need to synchronize, but might be wrong

File: pkg/cache/v3/status.go [hunk 19]
```diff
@@ -65,7 +66,9 @@ type statusInfo struct {
 	node *core.Node
 
 	// watches are indexed channels for the response watches and the original requests.
-	watches map[int64]ResponseWatch
+	// watches map[int64]ResponseWatch
```
Comment (valerian-roche) [inline pkg/cache/v3/status.go:69]: Minor: comment to cleanup?

File: pkg/server/sotw/v3/xds.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (valerian-roche) [reply pkg/server/sotw/v3/xds.go:97]: For now I think adding a comment about the risk might be useful.
Longer term I'd like to do more work on improving the state machine and aligning between delta and sotw, and that work will hopefully cover this issue

File: sample/bootstrap-xds.yaml [hunk 20]
```diff
@@ -53,16 +53,16 @@ static_resources:
...
-        rtds_config:
-          resource_api_version: V3
-          api_config_source:
-            transport_api_version: V3
-            api_type: GRPC
-            grpc_services:
-              envoy_grpc:
-                cluster_name: xds_cluster
-        name: runtime-0
+# layered_runtime:
```
Comment (valerian-roche) [inline sample/bootstrap-xds.yaml:56]: Is there a reason for this removal? Is there an issue with a new dependency version?

**Review by valerian-roche — APPROVED** [review]
Nothing blocking, just a comment for potential order issue would be nice

File: pkg/cache/v3/status.go (hunk 19 above)
Comment (alecholmez) [reply pkg/cache/v3/status.go:69]: Oops I'll clean that up

File: pkg/server/sotw/v3/xds.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/server/sotw/v3/xds.go:97]: Good call I can add a quick comment then we can get this merged

**Review by valerian-roche — APPROVED** [review]
